src/test/java/jenkins/plugins/logstash/PipelineTest.java

```diff
@@ -38,20 +39,28 @@ public void setup() throws Exception
       config.setEnabled(true);
   }
 
+  @Issue("JENKINS-61735")
   @Test
   public void logstash() throws Exception
   {
     WorkflowJob p = j.jenkins.createProject(WorkflowJob.class, "p");
-    p.setDefinition(new CpsFlowDefinition("logstash {\n" +
-        "currentBuild.result = 'SUCCESS'\n" +
-        "echo 'Message'\n" +
-    "}", true));
+    p.setDefinition(new CpsFlowDefinition(
+            "node('master') {\n" +
+            "  stage('mystage') {\n" +
+            "    logstash {\n" +
+            "      currentBuild.result = 'SUCCESS'\n" +
+            "      echo 'Message'\n" +
+            "    }\n" +
+            "  }\n" +
+            "}", true));
```

Comment:
I think this will only work then the logstash step is insside a node/stage. When you write
`logstash {
 stage('s') {
    echo 'm'
 }`

Will it work as well?

Reply:
If I do:
    p.setDefinition(new CpsFlowDefinition(
            "node('master') {\n" +
                    "  logstash {\n" +
                    "    stage('mystage') {\n" +
                    "      currentBuild.result = 'SUCCESS'\n" +
                    "      echo 'Message'\n" +
                    "    }\n" +
                    "  }\n" +
                    "}", true));
Then you will have the NODE_NAME but not the STAGE_NAME.
In fact, it's normal to get the NODE_NAME as in this case the node is evaluated before the logstash{} while the STAGE_NAME is not yet evaluated, so it cannot be set.
But in fact, it's an expected behavior as we recover the envVars in the context of the evaluation of logstash{}.
In my use case (e.g. having the capacity to have at least STAGE_NAME + NODE_NAME in metadata associated to each line of log), I can deal with that as it is only a matter of putting the logstash{} statement at the right place.
FYI, I've in mind another case where the pipeline fails at its very begining (e.g. Cannot recover the jenkinsfile from git. We have this issue frequently...), the only possibility I can see is to get the logs is to systematically create a "post { failure {logstashSend()}}" so that I'll have at least the very first line of logs for this kind of issue.

Reply:
@mwinter69 I was thinking about your remark. 
I've currently recovered the envVars from the context that is set in LogstashStep.java
I was wondering whether it was possible to recover the envVars from the context at each line of log level. 
And maybe if it was then possible to setup globally the logstash for a pipeline, then it would cover all my use cases.

Reply:
I'm just implementing the possibility to enable globally with a TaskListenerDecorator. Not sure if there is a way to get to the corresponding flow node so one can extract from there (via the parents) the node and stage. 
There is a completely other possibility to get the logs to ES by implementing a true pipeline logger (like in https://github.com/jenkinsci/pipeline-cloudwatch-logs-plugin). There it is possible to get the node and stage.
But this comes at a high price:
- You need to also implement reading from ES as the logs are no longer stored in the file system (so how to handle rabbitmq, redis,...,)
- performance is bad when writing directly to ES (might be necessary to use a buffer inbetween like fluentd). Haven't tested this but logstash plugin might have this bad performance impact as well

We have implemented this pipeline logger in our company, we might once publish this as open source, but it is still WIP. What we currently try to do is writing to ES and in parallel write to local Filesystem, so we don't need to read from ES.
Also our implemention is much more lightweigth in regarding the data we sent to ES (see https://issues.jenkins-ci.org/browse/JENKINS-54685) and uses a more straight forward data model

Reply:
@mwinter69 Thanks for your answer. Is it possible to share your implementation of enabling globally with a TaskListenerDecorator ?

Reply:
sure I just opened a PR #93

Reply:
I've included stage and agent name now as well using a different approach. Still limited to the logstash step only when inside a stage/node. I think this is the better approach as this information is now inside the data field and not somehow "hidden" in the build environment field

Reply:
We can also include the branch name of parallel steps if desired

Reply:
@mwinter69 I've tested your version of code by using logstash 2.3.2
I've setup in jenkins->manage jenkins the "enable globally" flag.
But I don't see any logs in my Graylog for a pipeline without any logstash{} statement inside.
And when I set logstash{} inside a steps{} then I come back to initial behavior, meaning that I do not see the stage name and the node name.
Maybe I miss something?

My pipeline with logstash{} looks like this:
`pipeline {
    agent { label 'master' }
    stages {
        stage('Stage in parallel') {
            parallel {
                stage('My stage 1') {
                    steps {
                        logstash {
                            sh '''#!/bin/bash
                                echo "$STAGE_NAME runs on $NODE_NAME"
                            '''
                        }
                    }
                }
                stage('My stage 2') {
                    steps {
                        logstash {
                            sh '''#!/bin/bash
                                echo "$STAGE_NAME runs on $NODE_NAME"
                            '''
                        }
                    }
                }
            }
        }
    }
}

My pipeline without logstash looks like this:
pipeline {
    agent { label 'master' }
    stages {
        stage('Stage in parallel') {
            parallel {
                stage('My stage 1') {
                    steps {
                            sh '''#!/bin/bash
                                echo "$STAGE_NAME runs on $NODE_NAME"
                            '''
                    }
                }
                stage('My stage 2') {
                    steps {
                            sh '''#!/bin/bash
                                echo "$STAGE_NAME runs on $NODE_NAME"
                            '''
                    }
                }
            }
        }
    }
}`

Reply:
version 2.3.2 doesn't contain my changes. They aren't yet released.
When using the global logger it is not possible to get the node and stage information. In my approach it will only collect the nearest surrounding node/stage and add this to the data.

Reply:
@mwinter69 Thanks for your response. 
Do you think that it is worth for me to continue working on this proposal, or it is better that I wait for your code to be released?

Reply:
@tgatinea  you should be able to get a pre-relase version of the plugin as described in https://jenkins.io/blog/2018/05/15/incremental-deployment/ (or you should be if the builds weren't failing all the time). 
If this fails you can just build the version locally.
Can you try it that way and see if it resolves your problem?

Reply:
@tgatinea bump
do you need some more help here?

Reply:
I've hit this same problem with pipelines not sending NODE_NAME and this pull request looked promising and it does now send NODE_NAME and STAGE_NAME, however the NODE_NAME is incorrect.

What I see when I run a build on a node that NODE_NAME is set to "master" when I'm expecting it to be "armv7" which is the name of the node it ran on. I double checked this by getting the pipeline to run `sh 'env >/tmp/test` to dump the environment to the slave & it does show NODE_NAME as "armv7".

What I think is happening is that the changes are returning the environment of the Jenkins master rather than the one on the node.

For me this is an issue as I run parallel builds on multiple CPU architectures so right now I'm seeing logs for different architectures merged together.

I'll do some more digging around as the information is there.

Reply:
It depends on where put the logstash step in the pipeline. The step will only detect the node and stage once at the beginning of the step. So when you have the node inside logstash 
`logstash {
  node('armv7') {
   ...
  }
}
`
you will get master,
if you have 
`node('armv7') {
   logstash {
   ...
  }
}
`
you should probably get the correct node. But this change will most likely not make it. #93 (which is merged to master) provides a different solution for this that is known to work with the same limitations as just mentioned.

The only solution to get the node and stage correct for each step is to implement the LogStorage interface. But doing so means you have to take care to also write the log file to Jenkins itself or you have to implement also a reader from ES so Jenkins is able to show a log.
You might want to look at https://github.com/SAP/elasticsearch-logs-plugin
This is doing exactly this. But it has a different message format and doesn't log things like tests, the environment and other things. The implementation was done to be able to persist the logs of a JenkinsFileRunner execution inside a docker container to ES.
As additional feature it sends events to ES for each step that is started or stopped.
Note that this plugin is not available on jenkins in the plugin update site. I would consider it as early beta state, it might change incompatibly in the future. It might in the future become an official Jenkins plugin and might also change name.

Reply:
> The only solution to get the node and stage correct for each step is to implement the LogStorage interface.

For a while I thought we could query the pieline build for current stage somehow, but for parallel builds there is no single current stage.

Reply:
Just an idea: Blueocean is able to present each step separately etc. I expect this data is annotated somewhere. Maybe it could be used to determine the stage?

Reply:
The data is not annotated. The FileLogStorage creates a side index file that contains the information about which bytes in the log belong to which step. This information is written directly when the logfile itself is written after it was processed by any of the TaskListenerDecorators (e.g. timestamper, logstash and maybe others). So no chance to get this.
As mentioned only by implementing the LogStorage you are able to get the stage and node information and attach this information to each log line.

Reply:
OK, so if that is true I think this PR should be closed, as there is no way the plugin can provide this information with current approach

Reply:
I come back after some test in our CI with this version of the plugin logstash
@peter-mount I did not see your issue and probably it is due to the fact that in my pipeline is as following:
```
agent {  label 'linux-channel-build' }
    steps {
        logstash {
            ...
        }
}
```
And with this kind of pipeline, the NODE_NAME and STAGE_NAME are correct.

Reply:
> Does it means that I cannot propagate this PR on your last version that you pushed in master ?

I think he meant that the issue it's attempting to solve is already handled. If you are still missing something from the current master then by all means let's continue the work on this

Reply:
> That's really annoying because I would expect that my pipeline would not fail even if an issue appears while logstash sends its logs.

I think this should be controled by the `failBuild` setting

Reply:
@jakub-bochenski What is the failBuild setting ? It is related to logstash in particular or it is a more general setting ?

Reply:
@jakub-bochenski Forget my question. You were tlking about the parameter failBuild that can be set into logstashSend()
I'm going to check how to set this parameter with logstash{} statement

Reply:
@jakub-bochenski I have tried to reproduce my issue in a sandbox with th following pipeline:
```
pipeline {
    agent { label 'master' }
    stages {
        stage('Stage in parallel') {
            parallel {
                stage('My stage 1') {
                    steps {
                        logstash {
                            sh '''#!/bin/bash
                            for i in {1..10000}
                            do
                               echo "#1 Welcome $i times"
                            done
                            '''
                        }
                    }
                }
                stage('My stage 2') {
                    steps {
                        logstash {
                            sh '''#!/bin/bash
                            for i in {1..10000}
                            do
                               echo "#2 Welcome $i times"
                            done
                            '''
                        }
                    }
                }
            }
        }
    }
}
``` 
But after a while I get a lot of errors in the jenkins log :
``̀`
2020-05-19 11:51:53.185+0000 [id=12]	WARNING	o.j.p.w.log.FileLogStorage#maybeFlush: failed to flush /var/jenkins_home/jobs/afac/builds/1/log
java.io.IOException: Stream Closed
	at java.io.FileOutputStream.writeBytes(Native Method)
	at java.io.FileOutputStream.write(FileOutputStream.java:326)
	at org.jenkinsci.plugins.workflow.log.DelayBufferedOutputStream$FlushControlledOutputStream.write(DelayBufferedOutputStream.java:125)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.FilterOutputStream.flush(FilterOutputStream.java:140)
	at org.jenkinsci.plugins.workflow.log.FileLogStorage.maybeFlush(FileLogStorage.java:190)
	at org.jenkinsci.plugins.workflow.log.FileLogStorage.overallLog(FileLogStorage.java:198)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.getLogText(WorkflowRun.java:1046)
	at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
	at org.kohsuke.stapler.Function$MethodFunction.invoke(Function.java:396)
	at org.kohsuke.stapler.Function$InstanceFunction.invoke(Function.java:408)
	at org.kohsuke.stapler.MetaClass$2.doDispatch(MetaClass.java:219)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:747)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:878)
	at org.kohsuke.stapler.MetaClass$9.dispatch(MetaClass.java:456)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:747)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:878)
	at org.kohsuke.stapler.MetaClass$4.doDispatch(MetaClass.java:280)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:747)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:878)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:676)
	at org.kohsuke.stapler.Stapler.service(Stapler.java:238)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:755)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1617)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:154)
	at org.jenkinsci.plugins.ssegateway.Endpoint$SSEListenChannelFilter.doFilter(Endpoint.java:248)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at jenkins.security.ResourceDomainFilter.doFilter(ResourceDomainFilter.java:76)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at io.jenkins.blueocean.auth.jwt.impl.JwtAuthenticationFilter.doFilter(JwtAuthenticationFilter.java:61)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at io.jenkins.blueocean.ResourceCacheControl.doFilter(ResourceCacheControl.java:134)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at net.bull.javamelody.MonitoringFilter.doFilter(MonitoringFilter.java:239)
	at net.bull.javamelody.MonitoringFilter.doFilter(MonitoringFilter.java:215)
	at net.bull.javamelody.PluginMonitoringFilter.doFilter(PluginMonitoringFilter.java:88)
	at org.jvnet.hudson.plugins.monitoring.HudsonMonitoringFilter.doFilter(HudsonMonitoringFilter.java:114)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at jenkins.metrics.impl.MetricsFilter.doFilter(MetricsFilter.java:125)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at org.jenkinsci.plugins.modernstatus.ModernStatusFilter.doFilter(ModernStatusFilter.java:52)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at jenkins.telemetry.impl.UserLanguages$AcceptLanguageFilter.doFilter(UserLanguages.java:128)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:157)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:153)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)
	at hudson.security.UnwrapSecurityExceptionFilter.doFilter(UnwrapSecurityExceptionFilter.java:51)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:118)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.providers.anonymous.AnonymousProcessingFilter.doFilter(AnonymousProcessingFilter.java:125)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.rememberme.RememberMeProcessingFilter.doFilter(RememberMeProcessingFilter.java:142)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.AbstractProcessingFilter.doFilter(AbstractProcessingFilter.java:271)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.BasicHeaderProcessor.doFilter(BasicHeaderProcessor.java:93)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.context.HttpSessionContextIntegrationFilter.doFilter(HttpSessionContextIntegrationFilter.java:249)
	at hudson.security.HttpSessionContextIntegrationFilter2.doFilter(HttpSessionContextIntegrationFilter2.java:67)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:90)
	at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:82)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at jenkins.security.SuspiciousRequestFilter.doFilter(SuspiciousRequestFilter.java:36)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:545)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:566)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1610)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1300)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:485)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1580)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1215)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:500)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:547)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
	at java.lang.Thread.run(Thread.java:748)

...

2020-05-19 11:52:02.119+0000 [id=1180]	INFO	o.j.p.w.s.concurrent.Timeout#lambda$ping$0: org.jenkinsci.plugins.workflow.steps.durable_task.DurableTaskStep [#1264]: checking /var/jenkins_home/workspace/afac on  unresponsive for 10 sec
2020-05-19 11:52:02.126+0000 [id=1180]	INFO	o.j.p.w.s.concurrent.Timeout#lambda$ping$0: org.jenkinsci.plugins.workflow.steps.durable_task.DurableTaskStep [#1265]: checking /var/jenkins_home/workspace/afac on  unresponsive for 10 sec
2020-05-19 11:52:07.120+0000 [id=1180]	INFO	o.j.p.w.s.concurrent.Timeout#lambda$ping$0: org.jenkinsci.plugins.workflow.steps.durable_task.DurableTaskStep [#1264]: checking /var/jenkins_home/workspace/afac on  unresponsive for 15 sec
2020-05-19 11:52:07.126+0000 [id=1180]	INFO	o.j.p.w.s.concurrent.Timeout#lambda$ping$0: org.jenkinsci.plugins.workflow.steps.durable_task.DurableTaskStep [#1265]: checking /var/jenkins_home/workspace/afac on  unresponsive for 15 sec
2020-05-19 11:52:12.121+0000 [id=1180]	INFO	o.j.p.w.s.concurrent.Timeout#lambda$ping$0: org.jenkinsci.plugins.workflow.steps.durable_task.DurableTaskStep [#1264]: checking /var/jenkins_home/workspace/afac on  unresponsive for 20 sec
2020-05-19 11:52:12.126+0000 [id=1180]	INFO	o.j.p.w.s.concurrent.Timeout#lambda$ping$0: org.jenkinsci.plugins.workflow.steps.durable_task.DurableTaskStep [#1265]: checking /var/jenkins_home/workspace/afac on  unresponsive for 20 sec
```

Reply:
sorry, I don't quite follow
if it's a separate issue from this pull request, please file it in the JIRA tracker/GH issue with full description

Reply:
OK. I will do that as in fact this issue is not related to my pull request

Reply:
I've written the JIRA issue https://issues.jenkins-ci.org/browse/JENKINS-62354 for that.